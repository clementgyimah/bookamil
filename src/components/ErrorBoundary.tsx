import * as React from 'react'
import '../assets/css/ErrorBoundary.css'
import '../assets/css/FontSize.css'
import { IErrorBoundaryProps, IErrorBoundaryState } from '../types'

class ErrorBoundary extends React.Component<IErrorBoundaryProps, IErrorBoundaryState> {
  constructor(props: IErrorBoundaryProps) {
    super(props)
    this.state = { error: null, errorInfo: null }
  }

  componentDidCatch(
    error: IErrorBoundaryState['error'],
    errorInfo: IErrorBoundaryState['errorInfo']
  ) {
    // Catch errors in any components below and re-render with error message
    this.setState({
      error,
      errorInfo,
    })
    // You can also log error messages to an error reporting service here
  }

  render() {
    if (this.state.errorInfo) {
      // Error path
      return (
        <div className="error-main-container">
          <div className="error-sub-container">
            <div className="title30 error-title-text">Something went wrong...</div>
            <details className="error-details-title">
              {this.state.error && this.state.error.toString()}
              <br />
              {this.state.errorInfo.componentStack}
            </details>
          </div>
        </div>
      )
    }
    // Normally, just render children
    return this.props.children
  }
}

export { ErrorBoundary }
