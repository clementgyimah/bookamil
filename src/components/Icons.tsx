import '../assets/css/Icon.css'
import { AiFillCloseCircle } from 'react-icons/ai'
import { IoOpenOutline } from 'react-icons/io5'
import { IIconProps } from '../types'

const AiIcon = (props: IIconProps) => {
  const handleIconAction = () => {
    if (props.iconAction) {
      props.iconAction()
    }
  }

  return (
    <AiFillCloseCircle
      color={props.color}
      size={props.size}
      title={props.title}
      className={props.className}
      onClick={handleIconAction}
    />
  )
}

const Io5Icon = (props: IIconProps) => {
  const handleIconAction = () => {
    if (props.iconAction) {
      props.iconAction()
    }
  }

  return (
    <IoOpenOutline
      color={props.color}
      size={props.size}
      title={props.title}
      className={props.className}
      onClick={handleIconAction}
    />
  )
}

export { AiIcon, Io5Icon }
