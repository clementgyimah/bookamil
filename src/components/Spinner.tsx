import '../assets/css/HomePage.css'
import '../assets/css/Animations.css'

const Spinner = () => {
  return (
    <div className="spinner-main-container">
      <div aria-label="book-loader" className="spinner-style spinner-animation">
        <span>{null}</span>
      </div>
    </div>
  )
}

export { Spinner }
