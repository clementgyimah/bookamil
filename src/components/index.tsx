export * from './BookCard'
export * from './BookDetailModal'
export * from './Icons'
export * from './Button'
export * from './Spinner'
export * from './ErrorBoundary'
