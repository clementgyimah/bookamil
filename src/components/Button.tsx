import '../assets/css/Button.css'
import '../assets/css/FontSize.css'
import { IButtonProps } from '../types'
import { Io5Icon } from './Icons'

const Button = (props: IButtonProps) => {
  return (
    <button className="button-style" onClick={props.buttonAction}>
      <span className="normal-text15" id="button-text">
        {props.title}
      </span>
      {props.showOpenNewTabIcon && (
        <span className="open-new-tab-icon-container">
          <Io5Icon color="white" size={15} title="open in new tab icon" />
        </span>
      )}
    </button>
  )
}

export { Button }
