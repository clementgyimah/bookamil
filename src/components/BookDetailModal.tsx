import '../assets/css/BookDetailModal.css'
import '../assets/css/FontSize.css'
import { IBookModalProps, TMouseEvent } from '../types'
import { AiIcon } from './Icons'
import { Button } from './Button'
import { textValidator } from '../functions'

const BookDetailModal = (props: IBookModalProps) => {
  const modalClass = props.show ? 'book-modal-container' : 'book-modal-container-disable'
  const closeModalHandler = () => {
    return props.close()
  }

  // prevent modal dark background click from affecting modal
  const preventClick = (e: TMouseEvent) => e.stopPropagation()

  // download book handler
  const downloadBookPDF = (e: React.ChangeEvent<HTMLInputElement>) => {
    e.preventDefault()
    // check if pdf is available before downloading
    if (props.data.accessInfo.pdf.isAvailable && props.data.accessInfo.pdf.downloadLink) {
      window.location.href = props.data.accessInfo.pdf.downloadLink
    } else {
      alert('Oops ðŸ˜”, PDF is not available for this book')
    }
  }

  // read book handler
  const readBook = (e: React.ChangeEvent<HTMLInputElement>) => {
    e.preventDefault()
    window.open(props.data.volumeInfo.previewLink)
  }

  return (
    <div className={modalClass} onClick={closeModalHandler}>
      <div onClick={preventClick} className="book-details-container">
        <div className="thumbnail-container">
          <img
            alt="cover_image"
            src={props.data.volumeInfo.imageLinks.thumbnail}
            className="thumbnail-style"
          />
        </div>
        <div className="details-container">
          <div className="title-container">
            <span className="title20" id="title-text">
              {textValidator({ theText: props.data.volumeInfo.title, maxLength: 40 })}
            </span>
            <span className="title20" id="close-modal-icon">
              <AiIcon
                color="#E05263"
                size={30}
                title="modal close icon"
                iconAction={closeModalHandler}
                className="close-book-modal-icon"
              />
            </span>
          </div>
          <div className="book-description-container">
            <span className="title15" id="subtitle-text">
              {textValidator({ theText: props.data.volumeInfo.subtitle, maxLength: 100 })}
            </span>
            <span className="normal-text15" id="book-description-text">
              {textValidator({ theText: props.data.volumeInfo.description, maxLength: 900 })}
            </span>
          </div>
          <div className="button-container">
            <div className="download-book-button-container">
              <Button
                title="Download PDF"
                buttonAction={downloadBookPDF}
                showOpenNewTabIcon={false}
              />
            </div>
            <div className="read-book-button-container">
              <Button title="Read Book" buttonAction={readBook} showOpenNewTabIcon={true} />
            </div>
          </div>
        </div>
      </div>
    </div>
  )
}

export { BookDetailModal }
