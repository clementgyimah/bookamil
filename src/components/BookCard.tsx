import * as React from 'react'
import '../assets/css/BookCard.css'
import { IBookCardProps } from '../types'
import { BookDetailModal } from './BookDetailModal'

const BookCard = (props: IBookCardProps) => {
  const [showModal, setShowModal] = React.useState(false)

  const openModal = () => {
    setShowModal(true)
  }
  const closeModal = () => {
    setShowModal(false)
  }
  return (
    <div className="book-card-container">
      <BookDetailModal show={showModal} close={closeModal} data={props.bookItem} />
      <img
        alt="bookCover"
        src={props.bookItem.volumeInfo.imageLinks.thumbnail}
        className="image-style"
        onClick={openModal}
      />
    </div>
  )
}

export { BookCard }
