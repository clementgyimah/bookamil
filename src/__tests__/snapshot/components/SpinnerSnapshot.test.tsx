import renderer from 'react-test-renderer'
import { Spinner } from '../../../components'

test('spinner component snapshot test', () => {
  const spinnerSnap = renderer.create(<Spinner />)
  const spinnerSnapTree = spinnerSnap.toJSON()
  expect(spinnerSnapTree).toMatchSnapshot()
})
