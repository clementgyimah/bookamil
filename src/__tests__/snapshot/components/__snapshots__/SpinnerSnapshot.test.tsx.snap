// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`spinner component snapshot test 1`] = `
<div
  className="spinner-main-container"
>
  <div
    aria-label="book-loader"
    className="spinner-style spinner-animation"
  >
    <span />
  </div>
</div>
`;
